services/repository.py

```diff
@@ -187,82 +191,84 @@ async def update_commit_from_provider_info(repository_service, commit):
-                pullid = res.groups()[0]
-                pullid = pullid
-                commit.branch = (await repository_service.get_pull_request(pullid))[
-                    "base"
-                ]["branch"]
+    author_info = git_commit["author"]
+    if not author_info.get("id"):
+        commit_author = None
         log.info(
-            "Updated commit with info from git provider",
+            "Not trying to set an author because it does not have an id",
             extra=dict(
-                repoid=commit.repoid,
+                author_info=author_info,
+                git_commit=git_commit,
                 commit=commit.commitid,
-                branch_value=commit.branch,
-                author_value=commit.author_id,
             ),
         )
+    else:
+        commit_author = get_or_create_author(
+            db_session,
+            commit.repository.service,
+            author_info["id"],
+            author_info["username"],
+            author_info["email"],
+            author_info["name"],
+        )
+
+    commit.parent_commit_id = await fetch_appropriate_parent_for_commit(
+        repository_service, commit, git_commit
+    )
+    commit.message = git_commit["message"]
+    commit.author = commit_author
+    commit.updatestamp = datetime.now()
+    commit.timestamp = git_commit["timestamp"]
+
+    # attempt to populate commit.pullid from repository_service if we don't have it
+    if not commit.pullid:
+        commit.pullid = await repository_service.find_pull_request(
+            commit=commitid, branch=commit.branch
+        )
+
+    # if our records or the call above returned a pullid, fetch it's details
+    if commit.pullid:
+        pull_details = await repository_service.get_pull_request(pullid=commit.pullid)
+        # There's a chance that the commit comes from a fork
+        # so we append the branch name with the fork slug
+        branch_name = pull_details["head"]["branch"]
+        # TODO: 'slug' is in a `.get` because currently only GitHub returns that info
+        if pull_details["head"].get("slug") != pull_details["base"].get("slug"):
+            branch_name = pull_details["head"]["slug"] + ":" + branch_name
+        commit.branch = branch_name
+        commit.merged = False
+    else:
+        possible_branches = await repository_service.get_best_effort_branches(
+            commit.commitid
+        )
+        if commit.repository.branch in possible_branches:
+            commit.merged = True
+            commit.branch = commit.repository.branch
+        else:
+            commit.merged = False
+
+    if commit.repository.service == "bitbucket":
+        res = merged_pull(git_commit["message"])
+        if res:
+            pullid = res.groups()[0]
+            if int(pullid) != int(commit.pullid):
+                pull_details = await repository_service.get_pull_request(pullid)
+            commit.branch = pull_details["base"]["branch"]
+
+    db_session.flush()
```

Comment:
Just for my owner understanding, I'm curious to how this flush helps with reducing the number of updates. I had assumed that we had to either flush (or commit, which does a flush first) for the update to "happen" in the transaction, so I might have the wrong mental model on how it works!

Reply:
The number of updates are reduced by moving code around.
Previously, some fields on the current `Commit` were written, and those were apparently implicitly flushed by sqalchemy as another function was querying `Commit`s.
Now the queries happen first, and all the field updates second, whereby the `UPDATE` should update all the fields at once.

The `flush` is just there to force the `UPDATE` to happen inside of this function, as previously, it would be delayed till later on nested in a completely different call chain.

Though to be quite honest, I haven’t verified this behavior locally. We will see this more clearly in tracing.